graf2d/gpadv7/src/TObjectDrawable.cxx

```diff
@@ -26,31 +29,124 @@
+   }
+}
+
+////////////////////////////////////////////////////////////////////
+/// Provide css type
+
+std::string TObjectDrawable::DetectCssType(const TObject *obj)
+{
+   if (!obj) return "tobject";
+
+   const char *clname = obj->ClassName();
    if (strncmp(clname, "TH3", 3) == 0) return "th3";
+   if (strncmp(clname, "TH2", 3) == 0) return "th2";
+   if ((strncmp(clname, "TH1", 3) == 0) || obj->InheritsFrom("TH1")) return "th1";
    if (strncmp(clname, "TGraph", 6) == 0) return "tgraph";
    if (strcmp(clname, "TLine") == 0) return "tline";
    if (strcmp(clname, "TBox") == 0) return "tbox";
    return "tobject";
 }
 
+////////////////////////////////////////////////////////////////////
+/// Constructor, clones TObject instance
+
+TObjectDrawable::TObjectDrawable(TObject &obj) : RDrawable(DetectCssType(&obj))
+{
+   fKind = kObject;
+   auto clone = obj.Clone();
+   CheckOwnership(clone);
+   fObj = std::shared_ptr<TObject>(clone);
+}
+
+////////////////////////////////////////////////////////////////////
+/// Constructor, clones TObject instance
+
+TObjectDrawable::TObjectDrawable(TObject &obj, const std::string &opt) : TObjectDrawable(obj)
+{
+   SetOpt(opt);
+}
+
+////////////////////////////////////////////////////////////////////
+/// Constructor, takes ownership over the object
+
+TObjectDrawable::TObjectDrawable(TObject *obj) : RDrawable(DetectCssType(obj))
+{
+   fKind = kObject;
+   CheckOwnership(obj);
+   fObj = std::shared_ptr<TObject>(obj);
+}
+
+////////////////////////////////////////////////////////////////////
+/// Constructor, takes ownership over the object
+
+TObjectDrawable::TObjectDrawable(TObject *obj, const std::string &opt) : TObjectDrawable(obj)
+{
+   SetOpt(opt);
+}
+
+////////////////////////////////////////////////////////////////////
+/// Constructor
+
+TObjectDrawable::TObjectDrawable(const std::shared_ptr<TObject> &obj) : RDrawable(DetectCssType(obj.get()))
+{
+   fKind = kObject;
+   CheckOwnership(fObj.get());
+   fObj = obj;
+}
+
+
+////////////////////////////////////////////////////////////////////
+/// Constructor
+
+TObjectDrawable::TObjectDrawable(const std::shared_ptr<TObject> &obj, const std::string &opt) : TObjectDrawable(obj)
+{
+   SetOpt(opt);
+}
+
+////////////////////////////////////////////////////////////////////
+/// Creates special kind for for palette or list of colors
+/// One can create persistent, which does not updated to actual values
+
+TObjectDrawable::TObjectDrawable(EKind kind, bool persistent) : RDrawable("tobject")
+{
+   fKind = kind;
+
+   if (persistent)
+      fObj = CreateSpecials(kind);
+}
+
+////////////////////////////////////////////////////////////////////
+/// Destructor
+
+TObjectDrawable::~TObjectDrawable() = default;
+
 ////////////////////////////////////////////////////////////////////
 /// Convert TColor to RGB string for using with SVG
 
 const char *TObjectDrawable::GetColorCode(TColor *col)
 {
-   static TString code;
+   static std::string code;
```

Comment:
Please return a `std::string` and get rid of this `static` variable, to make the code thread safe from the start / by design, even if we don't see a use-case now.